vllm/entrypoints/openai/rpc/client.py

```diff
@@ -242,47 +356,63 @@ async def generate(
 
                 # Stream back the results from the RPC Server.
                 while not finished:
+
                     message = await socket.recv()
                     request_output = cloudpickle.loads(message)
 
                     if isinstance(request_output, Exception):
                         # On exception, check if the server is still healthy.
                         # Use this to set the sync `is_running` and `errored`
                         # properties.
-                        try:
-                            await self.check_health()
-                        except Exception:
-                            self._errored = True
+
+                        if not self._errored:
+                            try:
+                                await self.check_health(socket=socket)
+                            except Exception as e:
+                                self._errored = True
+                                finished = True
+                                logger.exception(
+                                    'Error occurred during execution: %s',
+                                    repr(e))
+
                         # NB: do before raising here so that the flag is set
                         # by the time the caller receives this exception
                         raise request_output
 
                     finished = request_output.finished
                     yield request_output
-        finally:
-            if not finished:
-                await self.abort(request_id)
 
-    async def check_health(self) -> None:
+            finally:
+                # Request was canceled by the client + we broke out of the loop.
+                if not finished:
+                    # Call abort over the RPC.
+                    await self.abort(request_id)
+
+                    # Drain the socket. TODO: is this needed?
+                    # This helps to avoid hitting the HWM in the proxy when the
+                    # under heavy load, since there will be 1-3 messages from
+                    # the RPCServer after abort is called. The draining here
+                    # ensures that the RPCServer is done sending
+                    # messages to this connection before closing the socket.
+                    while await socket.poll(timeout=VLLM_ABORT_DRAIN_TIMEOUT_MS
+                                            ) != 0:
+                        message = await socket.recv()
+                        request_output = cloudpickle.loads(message)
+
+                        # If aborted, CancelledError is injected.
+                        if request_output == asyncio.exceptions.CancelledError:
```

Comment:
I noticed that you didn't change this per my prior comment... probably because the current code is what works :)

It _shouldn't_ work though ... I traced this to a bug which I think I introduced myself recently. Opening a PR to fix that now.

Reply:
https://github.com/vllm-project/vllm/pull/7698 .. just adding a test